include/ttmlir/Dialect/TTNN/IR/TTNNWorkarounds.h

```diff
@@ -0,0 +1,154 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef TTMLIR_DIALECT_TTNN_IR_TTNN_LAYOUT_OPERAND_WORKAROUNDS_H
+#define TTMLIR_DIALECT_TTNN_IR_TTNN_LAYOUT_OPERAND_WORKAROUNDS_H
+
+#include "ttmlir/Dialect/TTNN/IR/TTNNOpsAttrs.h"
+#include <memory>
+#include <optional>
+#include <vector>
+
+namespace mlir::tt::ttnn::wa {
+using TensorLayoutWorkaround = std::optional<Layout>;
+using TensorBufferTypeWorkaround = std::optional<BufferType>;
+using TensorMemoryLayoutWorkaround = std::optional<TensorMemoryLayout>;
+// Class that encapsulates operand workarounds.
+// It contains tensor layout, tensor buffer type and tensor memory layout
+// workarounds.
+class TTNNOperandWorkarounds {
+public:
+  // Default constructor with no workarounds.
+  TTNNOperandWorkarounds() {}
+
+  // Constructor that takes tensor layout, tensor buffer type and tensor memory
+  TTNNOperandWorkarounds(
+      TensorLayoutWorkaround tensorLayoutWorkaround,
+      TensorBufferTypeWorkaround tensorBufferTypeWorkaround,
+      TensorMemoryLayoutWorkaround tensorMemoryLayoutWorkaround)
+      : tensorLayoutWorkaround(tensorLayoutWorkaround),
+        tensorBufferTypeWorkaround(tensorBufferTypeWorkaround),
+        tensorMemoryLayoutWorkaround(tensorMemoryLayoutWorkaround) {}
+
+  // Returns a specific workaround.
+  template <typename T>
+  std::optional<T> getWorkaround() const {
```

Comment:
IMO we should get rid of the getters and setters and just make the data members public members of a struct:
```c++
struct TTNNOperandWorkarounds {
  TensorLayoutWorkaround tensorLayoutWorkaround;
  TensorBufferTypeWorkaround tensorBufferTypeWorkaround;
  TensorMemoryLayoutWorkaround tensorMemoryLayoutWorkaround;
};
```

Reply:
I spoke with Radenko offline yesterday, we agreed that the current implementaiton is good enough, as it gives us templated access in the pass later and makes code more clean and readable.

Reply:
It just feels a bit artificial to me to 1-1 tie the type to the variable.  Why couldn't we eventually have?
```c++
TensorLayoutWorkaround tensorLayoutWorkaroundInterleaved;
TensorLayoutWorkaround tensorLayoutWorkaroundSharded;
```

Reply:
Not sure that I follow last reply...

TensorLayoutWorkaround captures whether the workaround is RowMajor/Tile layout. TTNNOperandWorkarounds abastracts the workaround in the essence you are proposing,for example, you can create:
TTNNOperandWorkarounds layoutRowMajorInterleavedWorkaround;

Reply:
I'm just thinking that in the future there might be more nuanced operand workarounds that require multiple members of the same type.  My example is a bit contrived, where we'd want to apply a different tensor layout depending on the buffer type, but the second we have to do this the getters and setters we have written on this class will no longer work.

Reply:
We mentioned both of these issues last week:
1. It's not possible to choose more than 1 value per "dimension" (as you mention in your example with interleaved+sharded)
2. It's not possible to create conditional constraints across dimensions, e.g. if dtype is fp16, then only interleaved works, but for f32 only sharded, etc.

The idea was to get this simpler version in, then try to make it work for these more complex situations. That said, if you do have suggestions to make this current approach more adaptable for future updates, let us know!

Tagging @mbezuljTT for suggestions, as he has experience in this area :)

Reply:
I understand your perspective now. This initial implementation includes just three simple workaround arguments that we can specify for each operation, without any dependencies on one another, or any more complex configurations. I chose this approach to address the majority of the issues we are currently facing. However, I agree that when we encounter more complex situations in the future, we will need to support a better mechanism.

To avoid blocking progress on this right now, I suggest we proceed with the current implementation for now, if you agree. In the meantime, I can think of a solution that will allow us to express more complex dependencies between workarounds and that solution can be:
We can get rid of the getter, and  I propose that we move applyWorkarounds(Operation *op) method to TTNNOperandWorkarounds class. Then, from our workaround pass, we will have something like this:

```
struct TTNNOperandWorkaroundsResult
{
 Layout targetTensorLayout;
 BufferType targetTensorBufferType;
 TensorMemoryLayout targetTensorMemoryLayout;
};

TTNNOperandWorkarounds
{
TTNNOperandWorkaroundsResult applyWorkarounds(Operation *op){...}
}

... pass code ...
TTNNOperandWorkaroundsResult appliedWorkarounds = operandWorkaround.applyWorkarounds(op);
```

Inside TTNNOperandWorkarounds class and applyWorkarounds(...) method we can capture all the dependencies we want, we can go even further and develop some kind of workaround solver that would take a different set of workarounds that we define for different ops and instead of having three specified workaround types like now we can specify more complex dependencies, hence TTNNOperandWorkarounds can become something like this:
```
class TensorLayoutWorkaroundType: WorkaroundType<Layout>
{
applyWorkaround(Operation *op, TTNNOperandWorkaroundsResult& res)....
}

TTNNOperandWorkarounds
{
public:
...
TTNNOperandWorkaroundsResult applyWorkarounds(Operation *op){
TTNNOperandWorkaroundsResult result;
for(workaround: workarounds)
{
workaround.applyWorkaround(op, result);
}
}
...
private:
std::vector<WorkaroundType> workarounds;
}
```

What do you say about this plan and approach @nsmithtt?

Reply:
> However, I agree that when we encounter more complex situations in the future, we will need to support a better mechanism.

Agreed, I'm all for checking in a simpler version first, but we should land the initial code architecture that leaves the door open for more complex situations in the future vs landing something that only works for the simple case.

In general I think getters and setters on a value type is not a good design pattern, and further, getters/setters that are specialized based on the member's type doesn't seem to add any convenience.

IMO this is much more readable and idiomatic:
```c++
operandWorkarounds.targetTensorLayout = rmLayout;
```

Over this:
```c++
operandWorkarounds.setWorkaround(rmLayout);
```

The latter makes me wonder how it works and if it's doing what I think it's doing.


Regarding your proposal, I think we should keep the interface signature as is:
```c++
TTNNOperandsWorkarounds getOperandsWorkarounds();
```

Why opposition to making `TTNNOperandsWorkarounds` a simple struct?  This seems super simple and captures what we want:
```c++
struct TTNNOperandWorkarounds {
  // Tensor layout workaround.
  TensorLayoutWorkaround tensorLayoutWorkaround;
  // Tensor buffer type workaround.
  TensorBufferTypeWorkaround tensorBufferTypeWorkaround;
  // Tensor memory layout workaround.
  TensorMemoryLayoutWorkaround tensorMemoryLayoutWorkaround;

  TTNNOperandWorkarounds(... constructor...) ... {}

  // These seem nice to keep:
  bool operator==...;
  bool hasAnyWorkaround() const...;
};

TTNNOperandsWorkarounds AbsOp::getOperandsWorkarounds() {
  TTNNOperandWorkarounds wa;
  wa.tensorLayoutWorkaround = tiled;
  return wa;
}
```

Reply:
Generally speaking:
- `struct`/_value_: Implies it's just data, just like `int` it's safe to copy around, treat as an immutable value type, doesn't have an identity.
- `class`/_object_: Implies it has an identity, have to think twice about copying it, might be stateful and takes extra consideration.

Objects have more cognitive overhead and complexity so whenever it's possible air on using the former.  Swift has explicitly adopted these semantics as part of the language and I think that's a great decision.

Reply:
> Regarding your proposal, I think we should keep the interface signature as is:
> 
> ```c++
> TTNNOperandsWorkarounds getOperandsWorkarounds();
> ```

I would like to clarify that I am not suggesting a change to the op interface itself. Instead, my proposal focuses on how to implement a more complex declaration of workarounds and focus on changes in TTNNOperandWorkarounds class\struct.

> Generally speaking:
> 
> * `struct`/_value_: Implies it's just data, just like `int` it's safe to copy around, treat as an immutable value type, doesn't have an identity.
> * `class`/_object_: Implies it has an identity, have to think twice about copying it, might be stateful and takes extra consideration.
> 
> Objects have more cognitive overhead and complexity so whenever it's possible air on using the former. Swift has explicitly adopted these semantics as part of the language and I think that's a great decision.

Regarding this, I agree, I updated my code to use the struct instead. Please take a look, and thanks for the suggestion!

Reply:
@azecevicTT, all of the `TTNNOperandsWorkarounds` members are optional so shouldn't we just default construct them as null opts?  I.e. this pattern can be written as my example above reads:
```c++
TTNNOperandsWorkarounds AbsOp::getOperandsWorkarounds() {
  TTNNOperandWorkarounds wa; // default constructed, everything nullopt
  wa.tensorLayoutWorkaround = tiled;
  return wa;
}
```

Reply:
It can, but I will give you one contrieved (please ignore that some of this doesn't exist yet and some have different names), but close to real example. I find this
```
TTNNOperandWorkarounds()
    .setWorkaround(Layout::Tile)
    .setWorkaround(DataFormat::BFLOAT_16)
    .setWorkaround(ShardSpec::L1Sharded);
```
much more cleaner and uniform that this
```
TTNNOperandWorkarounds wa;
wa.tensorLayoutWorkaround = Layout::Tile;
wa.tensorDataFormatWorkaround = DataFormat::BFLOAT_16;
wa.tensorShardSpecWorkaround = ShardSpec::L1Sharded;
```
Additional bonus of `setWorkaround` is that it's polymorphic method, so if in the future we have a use case where you receive some type of workaround that you don't know exactly which type is it, you don't have to query the type on the callsite, that knowledge is built into `setWorkaround`. On top of that with public members you have committed to their names, which might become the problem in the future if some components of architecture change and respective names of workarounds have to change.

Reply:
> On top of that with public members you have committed to their names, which might become the problem in the future if some components of architecture change and respective names of workarounds have to change.

My only thinking is that this gets a little weird if you have multiple workarounds with the same type.  It quickly gets complicated when there are dependencies, i.e. you must use tile tensor layout if data format is block compressed, or row major tensor layout for some op when running with buffer type interleaved.

It's unclear to me how `setWorkaround` scales in these complicated cases, you'd potentially end up with code like:
```c++
TTNNOperandWorkarounds()
  .setWorkaround(Layout::Tile) // This one is ambiguous?
  .setWorkaround(InterleavedTileLayout(Layout::RowMajor))
  .setWorkaround(BlockCompressedTileLayout(Layout::Tile));
```

You'd need to introduce tag types `InterleavedTileLayout` and `BlockCompressedTileLayout` so you could make a polymorphic `setWorkaround` member work.

One additional minor benefit of using the members names is discoverability.  Maybe you just want to see which ops use `tensorLayoutWorkaround`, you could grep the codebase and find all of the ops that have opted into this WA.

Reply:
Alright, I think I understand your line of reasoning now. I agree that solution with `setWorkaround` scales very badly in these more complicated situations. Solution with public members scales better, but only marginally, in the sense that as long as the number of these interdependencies are relatively small it might be okay, though I guess it's the best what we have for now.